src/core/file.cc

```diff
@@ -992,18 +1019,54 @@ xfs_concurrency_from_kernel_version() {
     return 0;
 }
 
+static const std::string NVME_STR = "nvme";
+static const std::string NG_STR = "ng";
+
+bool
+is_nvme_device(std::string name) {
+  size_t pos = name.find(NVME_STR);
+  if (pos != std::string::npos) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+int
+open_ng_device_fd(std::string nvme_file_path) {
+  std::string ng_file_path = nvme_file_path;
+  size_t pos = ng_file_path.find(NVME_STR);
+  assert(pos != std::string::npos);
+  ng_file_path.replace(pos, NVME_STR.size(), NG_STR);
+  return ::open(ng_file_path.c_str(), O_RDWR);
+}
+
 future<shared_ptr<file_impl>>
-make_file_impl(int fd, file_open_options options, int flags) noexcept {
-    return engine().fstat(fd).then([fd, options = std::move(options), flags] (struct stat st) mutable {
+make_file_impl(int fd, file_open_options options, int flags, std::string name) noexcept {
+    return engine().fstat(fd).then([fd, options = std::move(options), flags, name = std::move(name)] (struct stat st) mutable {
         auto st_dev = st.st_dev;
 
         if (S_ISBLK(st.st_mode)) {
-            size_t block_size;
+            size_t block_size = 0;
             auto ret = ::ioctl(fd, BLKBSZGET, &block_size);
             if (ret == -1) {
                 return make_exception_future<shared_ptr<file_impl>>(
                         std::system_error(errno, std::system_category(), "ioctl(BLKBSZGET) failed"));
             }
+
+            uint32_t nsid = -1;
+            int ng_device_fd = -1;
+#ifdef SEASTAR_HAVE_NVME
+            if (is_nvme_device(name)) {
+                if (engine().support_append()) {
+                    nvme_get_nsid(fd, &nsid);
+                    ng_device_fd = open_ng_device_fd(name);
+                    if (ng_device_fd != -1) {
+                        return make_ready_future<shared_ptr<file_impl>>(make_shared<blockdev_file_impl>(fd, open_flags(flags), options, st.st_rdev, block_size, ng_device_fd ,nsid));
+                    }
```

Comment:
`} else {` -- the file is transparently transformed into a classical block device file without new append functionality. I _think_ there should at least be the API to tell the caller that this had happened. Or the new open flag telling that we do want append-enhanced file so that open would fail open() if it cannot be achieved.